torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -68,95 +69,78 @@ def foreach_all_gather(
     group: dist.ProcessGroup,
     use_uint8: bool,
 ) -> None:
     world_size = group.size()
-    flat_tensor_offset = 0  # [0, total_sharded_numel - 1]
-    for fsdp_param in fsdp_params:
-        param_shards: List[torch.Tensor] = []
-        padded_sharded_numel = fsdp_param.all_gather_input_numel
-        dtype_size = fsdp_param.unsharded_param_data_dtype.itemsize
-        for rank in range(world_size):
-            param_shard_start = rank * padded_sharded_numel
-            numel_in_shard = min(
-                fsdp_param._unsharded_size.numel() - param_shard_start,
-                padded_sharded_numel,
-            )
-            if use_uint8:
-                # Multiply by the dtype size to convert to uint8 numel
-                numel_in_shard *= dtype_size
-            if numel_in_shard > 0:
-                param_shard = flat_tensor_shards[rank].narrow(
-                    0, flat_tensor_offset, numel_in_shard
-                )
-                if use_uint8:
-                    # View back to the expected dtype before the cat
-                    param_shard = param_shard.view(
-                        fsdp_param.unsharded_param_data_dtype
-                    )
-                param_shards.append(param_shard)
-        offset_increment = padded_sharded_numel
-        if use_uint8:
-            offset_increment *= dtype_size
-        flat_tensor_offset += offset_increment
+    if use_uint8:
+        split_sizes_per_rank = [
+            fsdp_param.all_gather_input_numel
+            * fsdp_param.unsharded_param_data_dtype.itemsize
+            for fsdp_param in fsdp_params
+        ]
+    else:
+        split_sizes_per_rank = [
+            fsdp_param.all_gather_input_numel for fsdp_param in fsdp_params
+        ]
+    split_sizes = split_sizes_per_rank * world_size
+    splits = torch.split(all_gather_output, split_sizes, dim=0)
+    num_fsdp_params = len(fsdp_params)
+    all_param_shards: List[List[torch.Tensor]] = []
+    outs: List[torch.Tensor] = []
+    for fsdp_param_idx, fsdp_param in enumerate(fsdp_params):
+        param_shards: List[torch.Tensor] = [
+            splits[fsdp_param_idx + rank * num_fsdp_params]
+            for rank in range(world_size)
+        ]
         fsdp_param.alloc_unsharded_param()
-        # See [Note: ``.data`` Usage in Per-Parameter Sharding]
-        out = fsdp_param._unsharded_param_data[: fsdp_param._unsharded_size.numel()]
-        if fsdp_param.unsharded_param_data_dtype == torch.float8_e4m3fn:
-            # HACK: While `aten::cat` is not supported, view as uint8
-            param_shards = [t.view(torch.uint8) for t in param_shards]
-            torch.cat(param_shards, out=out.data.view(torch.uint8))
-            continue
-        with _unsafe_preserve_version_counter(out):
-            torch.cat(
-                param_shards,
-                out=out,
-            )
+        out = fsdp_param._unsharded_param_data
+        if use_uint8:
+            out = out.view(torch.uint8)
```

Comment:
This is a trick where instead of calling `param_shard.view(unsharded_param_data_dtype)` for each shard to view back to the real dtype before cat, we conversely view `out` only once to uint8.